### PR TITLE
TradeShop v2.2.0 - Extended TS Addon Framework (TSAF)

### PR DESCRIPTION
**Pull request by SparklingComet**
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

### Pull Request
This PR serves the discussion and the development of the TS Addon f
Framework (TSAF), which is featured in version 2.2.0.

For now, the TSAF has four events, which the plugin fires for addons to listen to:

* `PlayerTradeEvent` [cancellable] - fired before the transaction is completed
* `SuccessfulTradeEvent` [final] - fired after the transaction is completed
* `PlayerShopOpenEvent` [cancellable] - fired when someone is about to open their shop
* `PlayerShopCloseEvent` [cancellable] - fired when someone is about to to close their shop
* `PlayerShopCreateEvent` [cancellable] - fired when someone is about to create a shop
* `PlayerShopDestroyEvent` [cancellable] - fired when someone is about to destroy a shop
* `PlayerShopInventoryOpenEvent` [cancellable] - fired when someone is about to open a shop's inventory
* `HopperShopAccessEvent` [pseudo-cancellable] - fired when a hopper attempts to access the inventory of a shop
* `PlayerShopChangeEvent` [cancellable] - fired when a shop owner/manager is about to modify shop information

Notice that “failed events” **do not fire any of these events**. This design is intended to work this way.

... with hopefully more to come.

More information on the TSAF can be found online:
https://tradeshop.github.io/tsaf.html

### CODE REVIEW

File: src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:570]: Can this be handled in the new event since we are passing it the player and shop anyways?

File: src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:599]: same as above, can this happen in new event?

File: src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java:145]: Same as above...

File: src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:197]: Same as above, not sure on this one though

File: src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:265]: ^

**Review by KillerOfPie — CHANGES_REQUESTED** [review]
I think with the events i may have to rewrite a good chunk of code to properly utilize them

**Comment by SparklingComet** [comment]
Not sure if I understood the question. What I think you are asking is, why don’t we move the execution of the transaction between buyer and shop to another listener for our events. If you meant something else apologies,just let me know. 

They most certainly can be handled in a new event listener, however that is not how it is intended to work. The event is supposed ro be listened to, so that when it is fired, the addon can do stuff or even interfere with the trade. If it does, and the event is cancelled, TradeShop stops doing what it was doing. If not, it finishes up. Either way, it would make things weird to simply move execution to another listener because we would 

a) have less oversight of our code
b) possibly interfere with potential addons
c) not need to move stuff to a listener for an event thay we fired in the first place.

I do, however, intend to do some code refactoring before this PR is merged, so that our listeners look cleaner.

**Review by KillerOfPie — APPROVED** [review]

**Comment by KillerOfPie** [comment]
Ill re review this later, thinking about it more i was just being crazy and made perfect sense the way it was being done.

**Comment by SparklingComet** [comment]
Javadocs for the TSAF went up at https://tradeshop.github.io/docs

File: src/main/java/org/shanerx/tradeshop/framework/package-info.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/framework/package-info.java:3]: Is this typed correctly, reads a little weird.

File: README.md
Comment (KillerOfPie) [inline README.md:5]: Does this need to be re written now?

File: src/main/resources/plugin.yml
Comment (KillerOfPie) [inline src/main/resources/plugin.yml]: Do we want to call it an economy trading plugin? kinda make it sounds like it is built around using currencies...

**Review by KillerOfPie — APPROVED** [review]
Githubs review updates are really nice btw... 

Just a couple things on  wording but not code problems that I can see, all looks good.

We may want to add a debug command that shows connected plugins and commands(if we still have that functionality) as well as all versioning information...

File: README.md
Comment (SparklingComet) [reply README.md:5]: Why? It is still the fundamental concept behind it.

File: src/main/java/org/shanerx/tradeshop/framework/package-info.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/framework/package-info.java:3]: I will look it over

File: src/main/resources/plugin.yml
Comment (SparklingComet) [reply src/main/resources/plugin.yml]: Fair enough, I will change it back

File: README.md
Comment (KillerOfPie) [reply README.md:5]: You right, its was just the write info on the sign part, but that can still be done. so this is good.

File: src/main/resources/plugin.yml
Comment (KillerOfPie) [reply src/main/resources/plugin.yml]: i think just taking out economy, the rest is fine, also maybe change the website to tradeshop.github.io

**Comment by SparklingComet** [comment]
> We may want to add a debug command that shows connected plugins and commands(if we still have that functionality) as well as all versioning information...

 
I removed the thing that forced addons to hook into our plugin by “registering” themselves because I fundamentally changed the design of the TSAF as I went along. I believe the old design was quite the performance hog, whereas this one is much more lightweight. So, at the moment, we cannot see who is listening to the events unless we loop through all the plugins and check their listeners...

A debug command is a good idea though, I will add it to the project manager and make a PR about it at some point.

File: src/main/java/org/shanerx/tradeshop/framework/package-info.java
Comment (KillerOfPie) [reply src/main/java/org/shanerx/tradeshop/framework/package-info.java:3]: could not figure out why this feels so weird to read so im leaving it for now.

**Comment by SparklingComet** [comment]
@KillerOfPie I would like to merge this if we are done here.

**Comment by SparklingComet** [comment]
Merging PR.
The docs’ scope will be adjusted at a later time.

**Pull request merged**
